### PR TITLE
plan, tidb: support plan cache for SELECT statement

### PR DESCRIPTION
**Pull request by zz-jason**
main change:
1. add a simple lru cache, defined in **plan/cache/lru.go**
2. in function `func (s *session) Execute(sql string) ([]ast.RecordSet, error)`, we check the opportunity to use plan cache first
3. currently, we only add select statement into cache

### CODE REVIEW

**Comment by zz-jason** [comment]
@shenli WIP, benchmark result will be provided later.

**Comment by winoros** [comment]
When transaction changed from to read only to not read only. The plan will be changed since dirty table exists.

**Comment by zimulala** [comment]
Please fix CI @zz-jason

File: executor/compiler.go [hunk 1]
```diff
@@ -38,20 +38,29 @@ func (c *Compiler) Compile(ctx context.Context, node ast.StmtNode) (ast.Statemen
...
+	cacheable := plan.Cacheable(node)
+
 	p, err := plan.Optimize(ctx, node, is)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
+	_, isDual := p.(*plan.TableDual)
+	if isDual {
+		cacheable = false
```
Comment (winoros) [inline executor/compiler.go]: Why dual cannot be cached?

Comment (zz-jason) [reply executor/compiler.go]:
```
create table t(a bigint);
select * from t where exists (select * from t);
```

this query will be optimized to a `TableDual`

File: session.go [hunk 2]
```diff
@@ -649,55 +650,115 @@ func (s *session) SetProcessInfo(sql string) {
...
+
+			stmt := &executor.ExecStmt{
+				InfoSchema:  infoSchema,
+				Plan:        plan,
+				Expensive:   expensive,
+				CanBeCached: cacheable,
+				Text:        stmtNode.Text(),
+			}
+			s.SetValue(context.QueryString, stmt.OriginText())
+			if cache.PlanCacheEnabled && len(stmtNodes) == 1 {
```
Comment (XuHuaiyu) [inline session.go]: what if `len(stmtNodes) > 1` ?

File: plan/cacheable_checker.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+// Cacheable checks whether the input ast is cacheable.
+func Cacheable(node ast.Node) bool {
+	checker := cacheableChecker{
+		cacheable: true,
+	}
+	node.Accept(&checker)
+	return checker.cacheable
+}
+
+type cacheableChecker struct {
```
Comment (XuHuaiyu) [inline plan/cacheable_checker.go]: add some comments for the rules of this checker.

File: plan/cache/sharded_lru.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+)
+
+var (
+	// PlanCacheEnabled stores the global config "plan-cache-enabled".
+	PlanCacheEnabled bool
+	// PlanCacheShards stores the global config "plan-cache-shards".
+	PlanCacheShards int64 = 200
+	// PlanCacheCapacity stores the global config "plan-cache-capacity".
+	PlanCacheCapacity int64 = 1000
+	// GlobalPlanCache stores the global plan cache for every session in a tidb-server
```
Comment (XuHuaiyu) [inline plan/cache/sharded_lru.go]: add `.` at the end

**Comment by zz-jason** [comment]
@coocood PTAL

File: plan/cache/simple_lru_test.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+	for i := 0; i < 5; i++ {
+		keys[i] = newMockHashKey(int64(i))
+		vals[i] = int64(i)
+		lru.Put(keys[i], vals[i])
+	}
+	c.Assert(lru.size, Equals, lru.capacity)
+	c.Assert(lru.size, Equals, int64(3))
+
+	// test for nonexistence elements
```
Comment (XuHuaiyu) [inline plan/cache/simple_lru_test.go]: s/ nonexistence/ non-existent

File: plan/cache/simple_lru_test.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+	for i := 0; i < 2; i++ {
+		hash := string(keys[i].Hash())
+		element, exists := lru.elements[hash]
+		c.Assert(exists, IsFalse)
+		c.Assert(element, IsNil)
+	}
+
+	// test for in existence elements
```
Comment (XuHuaiyu) [inline plan/cache/simple_lru_test.go]: s/ in existence/ existent

File: plan/cacheable_checker.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+		return in, true
+	}
+	return in, false
+}
+
+func (checker *cacheableChecker) canExprBeCached(exprNode ast.ExprNode) bool {
+	switch expr := exprNode.(type) {
+	case *ast.VariableExpr, *ast.ExistsSubqueryExpr:
+		return false
+	case *ast.UnaryOperationExpr:
```
Comment (coocood) [inline plan/cacheable_checker.go]: This is not needed, `V` will be visited.

File: plan/cacheable_checker.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+// will not be cached currently.
+// NOTE: we can add more rules in the future.
+type cacheableChecker struct {
+	cacheable bool
+}
+
+// Enter implements Visitor interface.
+func (checker *cacheableChecker) Enter(in ast.Node) (out ast.Node, skipChildren bool) {
+	switch node := in.(type) {
+	case *ast.SelectStmt:
```
Comment (coocood) [inline plan/cacheable_checker.go]: We don't need to do this, every expression in `SelectStmt` will be visited.

File: plan/cache/simple_lru_test.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+	keys := make([]*mockCacheKey, 5)
+	vals := make([]int64, 5)
+
+	for i := 0; i < 5; i++ {
+		keys[i] = newMockHashKey(int64(i))
+		vals[i] = int64(i)
+		lru.Put(keys[i], vals[i])
+	}
+
+	// test for nonexistence elements
```
Comment (XuHuaiyu) [inline plan/cache/simple_lru_test.go]: ditto

File: plan/cacheable_checker.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+		_, isExistsSubQuery := expr.V.(*ast.ExistsSubqueryExpr)
+		if isExistsSubQuery {
+			return false
+		}
+	}
+	return true
+}
+
+// Leave implements Visitor interface.
+func (checker *cacheableChecker) Leave(in ast.Node) (out ast.Node, skipChildren bool) {
```
Comment (coocood) [inline plan/cacheable_checker.go]: Just switch the type of `in` here will do.

File: executor/compiler.go [hunk 11]
```diff
@@ -22,36 +22,30 @@ import (
...
 
-// Compile compiles an ast.StmtNode to an ast.Statement.
-// After preprocessed and validated, it will be optimized to a plan,
-// then wrappped to an adapter *statement as stmt.Statement.
-func (c *Compiler) Compile(ctx context.Context, node ast.StmtNode) (ast.Statement, error) {
-	is := GetInfoSchema(ctx)
-	if err := plan.Preprocess(node, is, ctx); err != nil {
-		return nil, errors.Trace(err)
+// Compile compiles an ast.StmtNode to a physical plan.
+func (c *Compiler) Compile(ctx context.Context, node ast.StmtNode) (is infoschema.InfoSchema, p plan.Plan, expensive bool, cacheable bool, err error) {
```
Comment (coocood) [inline executor/compiler.go]: This function has too many return values, how about just return `ExecStmt`?

File: session.go [hunk 12]
```diff
@@ -649,57 +650,110 @@ func (s *session) SetProcessInfo(sql string) {
...
+				log.Warnf("[%d] compile error:\n%v\n%s", connID, err, sql)
+				terror.Log(errors.Trace(s.RollbackTxn()))
+				return nil, errors.Trace(err)
+			}
+			sessionExecuteCompileDuration.Observe(time.Since(startTS).Seconds())
+
+			// Step3: Cache the physical plan if possiable.
+			if cache.PlanCacheEnabled && cacheable && len(stmtNodes) == 1 {
+				if _, isSelect := stmtNodes[0].(*ast.SelectStmt); isSelect {
+					cache.GlobalPlanCache.Put(cacheKey, cache.NewSQLCacheValue(stmtNode, plan, expensive))
```
Comment (coocood) [inline session.go]: `isSelect` is covered by `cacheable` already.

File: plan/cache/sharded_lru.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+	"sync"
+
+	"github.com/spaolacci/murmur3"
+)
+
+var (
+	// PlanCacheEnabled stores the global config "plan-cache-enabled".
+	PlanCacheEnabled bool
+	// PlanCacheShards stores the global config "plan-cache-shards".
+	PlanCacheShards int64 = 200
```
Comment (coocood) [inline plan/cache/sharded_lru.go]: Use a power of 2 value is more efficient.

File: plan/cache/key.go [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+	readOnly       bool // stores the current tidb-server status.
+
+	hash []byte
+}
+
+// Hash implements Key interface.
+func (key *sqlCacheKey) Hash() []byte {
+	if key.hash == nil {
+		var (
+			userBytes  = []byte(key.user)
```
Comment (coocood) [inline plan/cache/key.go]: We can use `hack.Slice` to avoid memory allocation.

**Comment by zz-jason** [comment]
@coocood @XuHuaiyu @lamxTyler PTAL

**Comment by XuHuaiyu** [comment]
LGTM

File: plan/cacheable_checker.go [hunk 15]
```diff
@@ -37,42 +40,15 @@ type cacheableChecker struct {
...
-				return in, true
-			}
-		}
-		if !checker.canExprBeCached(node.Where) {
-			checker.cacheable = false
-			return in, true
-		}
-	case *ast.ExistsSubqueryExpr:
+	switch in.(type) {
+	case *ast.VariableExpr, *ast.ExistsSubqueryExpr:
```
Comment (coocood) [inline plan/cacheable_checker.go]: Query contains non-deterministic function like `Now()`, `RAND()` should not be cacheable.

File: plan/cacheable_checker.go [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/ast"
+)
+
+// Cacheable checks whether the input ast is cacheable.
+func Cacheable(node ast.Node) bool {
```
Comment (coocood) [inline plan/cacheable_checker.go]: Need some test for this function.

File: plan/cacheable_checker.go [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+// will not be cached currently.
+// NOTE: we can add more rules in the future.
+type cacheableChecker struct {
+	cacheable bool
+}
+
+// Enter implements Visitor interface.
+func (checker *cacheableChecker) Enter(in ast.Node) (out ast.Node, skipChildren bool) {
+	switch in.(type) {
+	case *ast.VariableExpr, *ast.ExistsSubqueryExpr:
```
Comment (coocood) [inline plan/cacheable_checker.go]: Query contains non-deterministic function like Now(), RAND() should not be cacheable.

File: session.go [hunk 18]
```diff
@@ -649,57 +650,102 @@ func (s *session) SetProcessInfo(sql string) {
...
+			executor.ResetStmtCtx(s, stmtNode)
+			stmt, err := compiler.Compile(s, stmtNode)
+			if err != nil {
+				log.Warnf("[%d] compile error:\n%v\n%s", connID, err, sql)
+				terror.Log(errors.Trace(s.RollbackTxn()))
+				return nil, errors.Trace(err)
+			}
+			sessionExecuteCompileDuration.Observe(time.Since(startTS).Seconds())
+
+			// Step3: Cache the physical plan if possiable.
```
Comment (coocood) [inline session.go]: possible

File: config/config.go [hunk 19]
```diff
@@ -121,6 +129,11 @@ var defaultConf = Config{
 		XHost: "0.0.0.0",
 		XPort: 14000,
 	},
+	PlanCache: PlanCache{
+		Enabled:  true,
```
Comment (coocood) [inline config/config.go]: Set default to true is not safe for now.

Comment (zz-jason) [reply config/config.go]: only for the ci test, will be disabled in the last commit of this pr

File: plan/cacheable_checker.go (hunk 16 above)
Comment (zz-jason) [reply plan/cacheable_checker.go]: added

File: plan/cacheable_checker.go (hunk 17 above)
Comment (zz-jason) [reply plan/cacheable_checker.go]: done. btw, rand() can be cached

File: session.go (hunk 18 above)
Comment (zz-jason) [reply session.go]: done

**Comment by coocood** [comment]
LGTM

**Pull request merged**